internal/balancergroup/balancergroup.go

```diff
@@ -182,6 +182,29 @@ func (sbc *subBalancerWrapper) stopBalancer() {
 	sbc.balancer = nil
 }
 
+type balancerCache interface {
+	// Add adds item with key to the cache. callback is invoked when item is
+	// removed from the cache.
+	Add(key, item interface{}, callback func()) (interface{}, bool)
+
+	// Remove the item with the key from the cache.
+	Remove(key interface{}) (item interface{}, ok bool)
+
+	// Clear removes all entries, and runs the callbacks if runCallback is true.
+	Clear(runCallback bool)
+}
+
+// A no-op implementation of the balancerCache interface when caching is
+// disabled due to Options.SubBalancerCloseTimeout set to 0.
+type noopBalancerCache struct{}
+
+func (noopBalancerCache) Add(_ interface{}, _ interface{}, cb func()) (interface{}, bool) {
+	cb()
```

Comment:
Executing this callback synchronously could theoretically cause problems with a pattern like:

```go
b.Lock()
b.cache.Add(blah, blah, func() { b.Lock(); do something; b.Unlock(); })
b.Unlock()
```

I wonder if we should just `go cb()` now while seeing this, or wait until we run into that problem, which may never happen anyway.

Reply:
I agree `go cb()` should have been what the noop implementation should have done. The problem probably doesn't happen with the current implementation of the timeout cache because the timer callback is invoked in a separate goroutine even if the timeout value is `0`.